src/Composer/Satis/Builder/PackagesBuilder.php

```diff
@@ -126,29 +134,71 @@ private function pruneIncludeDirectories()
     private function dumpPackageIncludeJson(array $packages, $includesUrl, $hashAlgorithm = 'sha1')
     {
         $filename = str_replace('%hash%', 'prep', $includesUrl);
-        $path = $this->outputDir . '/' . ltrim($filename, '/');
+        $path = $tmpPath = $this->outputDir . '/' . ltrim($filename, '/');
 
         $repoJson = new JsonFile($path);
-        $repoJson->write(array('packages' => $packages));
+        $contents = $repoJson->encode(array('packages' => $packages)) . "\n";
 
-        $hash = hash_file($hashAlgorithm, $path);
+        $hash = hash($hashAlgorithm, $contents);
 
         if (strpos($includesUrl, '%hash%') !== false) {
             $this->writtenIncludeJsons[] = array($hash, $includesUrl);
             $filename = str_replace('%hash%', $hash, $includesUrl);
-            rename(
-                $path,
-                $path = $this->outputDir . '/' . ltrim($filename, '/')
-            );
+            if (file_exists($path = $this->outputDir . '/' . ltrim($filename, '/'))) {
+                // When the file exists, we don't need to override it as we assume,
+                // the contents satisfy the hash
+                $path = null;
+            }
+        }
+        if ($path) {
+            $this->writeToFile($path, $contents);
+            $this->output->writeln("<info>wrote packages to $path</info>");
         }
-        $this->output->writeln("<info>wrote packages to $path</info>");
 
         return array(
             $filename => array($hashAlgorithm => $hash)
         );
     }
 
     /**
+     * Write to a file
+     *
+     * @param string $path
+     * @param string $contents
+     */
+    private function writeToFile($path, $contents)
+    {
+        $dir = dirname($path);
+        if (!is_dir($dir)) {
+            if (file_exists($dir)) {
+                throw new \UnexpectedValueException(
+                    $dir.' exists and is not a directory.'
+                );
+            }
+            if (!@mkdir($dir, 0777, true)) {
+                throw new \UnexpectedValueException(
+                    $dir.' does not exist and could not be created.'
+                );
+            }
+        }
+
+        $retries = 3;
+        while ($retries--) {
+            try {
+                file_put_contents($path, $contents);
+                break;
+            } catch (\Exception $e) {
```

Comment:
It isn't: file_put_contents not only returns false on error but triggers a warning - which will result in an error exception, which is just handled here:

    $> php -a
    file_put_contents("/test" ,"test");
    PHP Warning:  file_put_contents(/test): failed to open stream: Permission denied in php shell code on line 1